### PR TITLE
Update TemurinGenSBOM.java file (using CycloneDX core java to generate SBOM)

### PR DESCRIPTION
**Pull request by SehrishHussain**
Update TemurinGenSBOM.java

### CODE REVIEW

**Comment by SehrishHussain** [comment]
@andrew-m-leonard A new PR please take a look. Some Linter code errors I need to discuss.

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 1]
```diff
@@ -17,53 +15,234 @@
...
+        comp.setType(Component.Type.APPLICATION);
+        comp.setName(name);
+        bom.addComponent(comp);
+        return bom;
+    }
+    static Bom addComponentProperty(Bom bom, String name, String value) {     //Method to store Component->Property->name-value pairs
+        prop1.setName(name);
+        prop1.setValue(value);
+        prop.add(prop1);
+        comp.setProperties(prop);
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:194]: Not sure this comp gets persisted, as it's not set in the bom any where?

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 2]
```diff
@@ -17,53 +15,234 @@
...
+        bom.addProperty(prop1);
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:195]: This property is on the top level BOM, not the Component? need to identify which component?

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 3]
```diff
@@ -17,53 +15,234 @@
...
+        return bom;
+    }
+    static Bom addExternalReference(Bom bom, String url, String hashes, String comment) {   //Method to store externalReferences: dependency_version_alsa
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:198]: hashes is not used in the method?

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 4]
```diff
@@ -17,53 +15,234 @@
...
+        extRef.setHashes(hash);
+        extRef.addHash(hash1);
+        extRef.setUrl(url);
+        extRef.setComment(comment);
+        extRef.setType(ExternalReference.Type.BUILD_SYSTEM);
+        bom.addExternalReference(extRef);
+        return bom;
+    }
 
+    static Bom addComExternalReference(Bom bom, String url, String hashes, String comment) {  //Method to store externalReferences to store: openjdk_source
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:209]: addComp <- spelling

Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:209]: hashes not used

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 5]
```diff
@@ -17,53 +15,234 @@
...
+                comment = args[++i];
+            } else if (args[i].equals("--createNewSBOM")) {
+                cmd = "createNewSBOM";
+            } else if (args[i].equals("--addMetadata")) {        //This is Metadata Component. We can set "name" for Metadata->Component.
+                cmd = "addMetadata";
+            } else if (args[i].equals("--addMetadataProp")) {   //This is MetaData Component --> Property -> name-value: os, arch, variant, scmRef, buildRef, full_version_output, makejdk_any_platform_args
+                cmd = "addMetadataProperty";
+            } else if (args[i].equals("--addComponent")) {      //This is Components->Property: will add name-value.
+                cmd = "addComponent";
+            } else if (args[i].equals("--addComponentProp")) {  //This is Components->Property: will add name-value.
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:83]: You need to identify which Component Prop is being added to? maybe have a position parameter to the --addComponentProp arg which is the Component "name" to add prop to, eg:
--addComponentProp "thiscomp"

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 6]
```diff
@@ -17,53 +15,234 @@
...
+            } else if (args[i].equals("--value")) {
+                value = args[++i];
+            } else if (args[i].equals("--url")) {
+                url = args[++i];
+            } else if (args[i].equals("--hashes")) {
+                hashes = args[++i];
+            } else if (args[i].equals("--comment")) {
+                comment = args[++i];
+            } else if (args[i].equals("--createNewSBOM")) {
+                cmd = "createNewSBOM";
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:76]: Need validation that no other cmd has been specified, ie.if (cmd != null) { System.out.println("Syntax error: only one command can be specified..."); ...}

Comment (andrew-m-leonard) [reply cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:76]: this check probably needs adding to all args... eg. "--hashes already specified"

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 7]
```diff
@@ -17,53 +15,234 @@
...
+                bom = addComponent(bom, name);
+                String json = generateBomJson(bom);
+                writeJSONfile(json, fileName);
+                System.out.println("SBOM: " + json);
+            } break;
 
-        System.out.println("SBOM: " + json);
+            case "addComponentProp": {                             //This is Components adds name-value pairs to List
+                Bom bom = readJSONfile();
+                bom = addComponentProperty(bom, name, value);
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:125]: needs a "compRef" param for which Component to add prop to..

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 8]
```diff
@@ -17,53 +15,234 @@
...
+                cmd = "addMetadata";
+            } else if (args[i].equals("--addMetadataProp")) {   //This is MetaData Component --> Property -> name-value: os, arch, variant, scmRef, buildRef, full_version_output, makejdk_any_platform_args
+                cmd = "addMetadataProperty";
+            } else if (args[i].equals("--addComponent")) {      //This is Components->Property: will add name-value.
+                cmd = "addComponent";
+            } else if (args[i].equals("--addComponentProp")) {  //This is Components->Property: will add name-value.
+                cmd = "addComponentProp";
+            } else if (args[i].equals("--addExternalRef")) {
+                cmd = "addExtRef";
+            } else if (args[i].equals("--addComExtRef")) {
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:87]: use --addComponentExtRef, to be consistent with --addComponentProp etc..

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java (hunk 5 above)
Comment (SehrishHussain) [reply cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:83]: @andrew-m-leonard I have tried making multiple Components by constructing multiple objects, but they all seem to add in same-one component. So far there is only one Metadata and one Component, further, both have different attributes.

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java (hunk 7 above)
Comment (SehrishHussain) [reply cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:125]: @andrew-m-leonard I will work on this again to see if we can add multiple Components.

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java (hunk 1 above)
Comment (SehrishHussain) [reply cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:194]: @andrew-m-leonard I'm going to make a list of components and see if we can add multiple components that way.

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java (hunk 5 above)
Comment (SehrishHussain) [reply cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:83]: @andrew-m-leonard yes, working on it.

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java (hunk 7 above)
Comment (andrew-m-leonard) [reply cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:125]: @SehrishHussain don't worry about multiple Components in this PR

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java (hunk 8 above)
Comment (andrew-m-leonard) [reply cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:87]: don't worry about addComponentExtRef in this PR

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java (hunk 5 above)
Comment (andrew-m-leonard) [reply cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:83]: don't worry about Props in this PR

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 9]
```diff
@@ -17,53 +15,229 @@
...
 
+    static Bom createBom(final String name, final String version) {        //Create SBOM, test.JSON file
+        bom = new Bom();
+        System.out.println(bom.getBomFormat());
+        comp.setName(name);
+        comp.setVersion(version);
+        comp.setType(Component.Type.APPLICATION);
+        comp.setGroup("Eclipse Temurin");
+        comp.setAuthor("Vendor: Adoptium");
+        bom.addComponent(comp);
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:157]: Not sure what this comp is being added here?

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 10]
```diff
@@ -17,53 +15,229 @@
...
+    static Bom addComponentProperty(final Bom bom, final String name, final String value) {     //Method to store Component-> Property-> name-value pairs
+        prop1.setName(name);
+        prop1.setValue(value);
+        prop.add(prop1);
+        comp.setProperties(prop);
+        bom.addProperty(prop1);
+        return bom;
+    }
+    static Bom addExternalReference(final Bom bom, final String url, final String comment) {   //Method to store externalReferences: dependency_version_alsa
+        hash.add(hash1);
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:195]: and here:
```
ExternalReference extRef = new ExternalReference();
```
Remove the static global variables.
The "art" of good class structure is "encapsulation", eg.creating variables within scope for the purpose of just that method.

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java (hunk 9 above)
Comment (SehrishHussain) [reply cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:157]: Added this comp to show below data on top. 
    {
      "author" : "Vendor: Adoptium",
      "group" : "Eclipse Temurin",
      "name" : "Eclipse Temurin",
      "version" : "jdk11",
      "type" : "application"
    },
Now once we make multiple components, will move this data to top-level component.

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 11]
```diff
@@ -17,53 +15,232 @@
...
+            } else if (args[i].equals("--name")) {
+                name = args[++i];
+            } else if (args[i].equals("--value")) {
+                value = args[++i];
+            } else if (args[i].equals("--url")) {
+                url = args[++i];
+            } else if (args[i].equals("--comment")) {
+                comment = args[++i];
+            } else if (args[i].equals("--hashes")) {
+                hashes = arg[++i];
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:59]: should be args[++i]

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 12]
```diff
@@ -17,53 +15,232 @@
...
+        hash.add(hash1);
+        extRef.add(hash1);
+        extRef.setUrl(url);
+        extRef.setComment(comment);
+        extRef.setType(ExternalReference.Type.BUILD_SYSTEM);
+        bom.addExternalReference(extRef);
+        return bom;
+    }
 
+    static Bom addComExternalReference(final Bom bom, final String hashes, final String url, final String comment) {  //Method to store externalReferences to store: openjdk_source
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:203]: should be addComponentExternalReference(...

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 13]
```diff
@@ -17,53 +15,232 @@
...
+        prop1.setValue(value);
+        prop.add(prop1);
+        comp.setProperties(prop);
+        bom.addProperty(prop1);
+        return bom;
+    }
+    static Bom addExternalReference(final Bom bom, final String hashes, final String url, final String comment) {   //Method to store externalReferences: dependency_version_alsa
+        ExternalReference extRef = new ExternalReference();
+        Hash hash1 = new Hash(Hash.Algorithm.SHA3_256, hashes);
+        hash.add(hash1);
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:194]: hash.add(hash1); line not needed fails to compile, delete line

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 14]
```diff
@@ -17,53 +15,232 @@
...
+        extRef.add(hash1);
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:195]: should be extRef.addHash(hash1);

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 15]
```diff
@@ -17,53 +15,232 @@
...
+        bom.addExternalReference(extRef);
+        return bom;
+    }
 
+    static Bom addComExternalReference(final Bom bom, final String hashes, final String url, final String comment) {  //Method to store externalReferences to store: openjdk_source
+        ExternalReference extRef = new ExternalReference();
+        Hash hash1 = new Hash(Hash.Algorithm.SHA3_256, hashes);
+        Component comp = new Component();
+        hash.add(hash1);
+        extRef.setHashes(hash);
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:208]: remove following lines, as addHash(hash1) method does this below:
```
hash.add(hash1);
extRef.setHashes(hash);
```

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 16]
```diff
@@ -17,53 +15,232 @@
...
+            } break;
+            default: {
+                System.out.println("Please enter a command.");
+            }
+        }
     }
 
-    static Bom createTestBom() {
+    static Bom createBom(final String name, final String version) {        //Create SBOM, test.JSON file
+        System.out.println(bom.getBomFormat());
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:138]: this debug System.out.. line needs removing, as does not compile, bom not in scope.

**Review by andrew-m-leonard — CHANGES_REQUESTED** [review]

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 17]
```diff
@@ -17,53 +15,228 @@
...
+        Hash hash1 = new Hash(Hash.Algorithm.SHA3_256, hashes);
+        extRef.addHash(hash1);
+        extRef.setUrl(url);
+        extRef.setComment(comment);
+        extRef.setType(ExternalReference.Type.BUILD_SYSTEM);
+        bom.addExternalReference(extRef);
+        return bom;
+    }
 
+    static Bom addComExternalReference(final Bom bom, final String hashes, final String url, final String comment) {  //Method to store externalReferences to store: openjdk_source
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:201]: Should be addComponentExternalReference(..

File: cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java [hunk 18]
```diff
@@ -18,52 +17,223 @@
...
+                writeJSONfile(bom, fileName);
+                break;
+
+            case "addComponentProp":                       // Adds Components --> name-value pairs
+                bom = addComponentProperty(name, value, fileName);
+                writeJSONfile(bom, fileName);
+                break;
+
+            case "addExternalReference":                                     // Adds external Reference
+                bom = addExternalReference(url, hashes, comment, fileName);
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/src/temurin/sbom/TemurinGenSBOM.java:113]: url, hashes params are in different order to the addExternalReference() method below...

**Review by andrew-m-leonard — CHANGES_REQUESTED** [review]

File: cyclonedx-lib/build.xml [hunk 19]
```diff
@@ -107,8 +108,136 @@
...
+                </java> 
+              
+               <java classpath="${classpath}" classname="temurin.sbom.TemurinGenSBOM">
+                  <arg value="--addComponent"/>
+                  <arg value="--compName"/>
+                  <arg value="ScmRef"/>
+                  <arg value="--name"/>
+                  <arg value="scmRef"/>
+                  <arg value="--value"/>
+                  <arg value="dk-17+35_adopt"/>
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/build.xml:129]: to be consistent should be: "jdk-17+35_adopt"

File: cyclonedx-lib/build.xml [hunk 20]
```diff
@@ -18,7 +18,7 @@
         <taskdef resource="net/sf/antcontrib/antlib.xml" />
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/build.xml:18]: Add this at the beginning of this file, so Linter jscpd ignores this file
```
<!-- jscpd:ignore-start -->
```

File: cyclonedx-lib/build.xml [hunk 21]
```diff
@@ -130,4 +259,3 @@
 		</sequential>
 	</macrodef>
 </project>
```
Comment (andrew-m-leonard) [inline cyclonedx-lib/build.xml:261]: Add this at the end of this file, so Linter jscpd ignores this file
```
<!-- jscpd:ignore-end -->
```

**Review by andrew-m-leonard — CHANGES_REQUESTED** [review]

File: cyclonedx-lib/build.xml (hunk 21 above)
Comment (SehrishHussain) [reply cyclonedx-lib/build.xml:263]: @andrew-m-leonard added the lines in the correct place?

Comment (sxa) [reply cyclonedx-lib/build.xml:263]: Still looks like it's failing checks, so I'd imagine it needs to be the final line in the file after `</project>`.

Comment (SehrishHussain) [reply cyclonedx-lib/build.xml:263]: @sxa Still failing the check tho.

**Comment by SehrishHussain** [comment]
For reference, pasting the content of sample json file generated here:
 ```
{
  "bomFormat" : "CycloneDX",
  "specVersion" : "1.3",
  "version" : 1,
  "metadata" : {
    "timestamp" : "2022-02-24T10:48:35Z",
    "component" : {
      "type" : "application"
    },
    "manufacture" : {
      "name" : "Eclipse Foundation",
      "url" : [
        "https://www.eclipse.org/"
      ]
    },
    "properties" : [
      {
        "name" : "OS",
        "value" : "Linux"
      },
      {
        "name" : "arch",
        "value" : "x64"
      },
      {
        "name" : "variant",
        "value" : "hotspot"
      }
    ]
  },
  "components" : [
    {
      "author" : "Vendor: Adoptium",
      "group" : "Eclipse Temurin",
      "name" : "Temurin",
      "version" : "jdk17+35",
      "type" : "application"
    },
    {
      "name" : "ScmRef",
      "description" : "Adoptium openjdk mirror git tag",
      "properties" : [
        {
          "name" : "scmRef",
          "value" : "jdk-17+35_adopt"
        }
      ]
    },
    {
      "name" : "buildRef",
      "description" : "git commit of temurin-build script",
      "properties" : [
        {
          "name" : "buildRef",
          "value" : "https://github.com/adoptium/temurin-build/commit/c3a40"
        }
      ]
    },
    {
      "name" : "full_version_output",
      "properties" : [
        {
          "name" : "full_version_output",
          "value" : "openjdk version: 17 2021-09-14\\nOpenJDK Runtime Environment Temurin-17+35 (build 17+35)\\nOpenJDK 64-Bit Server VM Temurin-17+35 (build 17+35, mixed mode, sharing)\\n"
        }
      ]
    },
    {
      "name" : "makejdk_any_platform",
      "description" : "temurin build make arguments",
      "properties" : [
        {
          "name" : "makejdk_any_platform_args",
          "value" : "--clean-git-repo --jdk-boot-dir /usr/lib/jvm/jdk-16 --configure-args --disable-warnings-as-errors --enable-ccache --enable-dtrace --target-file-name OpenJDK17-jdk_x64_linux_hotspot_17_35.tar.gz --release --clean-libs --tag jdk-17+35_adopt --skip-freetype --use-jep319-certs --create-debug-image --build-variant hotspot jdk17"
        }
      ]
    },
    {
      "name" : "make_command_args",
      "properties" : [
        {
          "name" : "make_command_args",
          "value" : "make product-images legacy-jre-image test-image"
        }
      ]
    }
  ],
  "externalReferences" : [
    {
      "type" : "build-system",
      "url" : "https://github.com/adoptium/jdk17/commit/a5afad28437",
      "comment" : "openjdk_source"
    },
    {
      "type" : "build-system",
      "url" : "https://ftp.osuosl.org/pub/blfs/conglomeration/alsa-lib/alsa-lib-1.1.6.tar.bz2",
      "comment" : "dependency_version_alsa"
    }
  ]
}

```

**Review by andrew-m-leonard — APPROVED** [review]
@SehrishHussain I'm happy with this now.

The Components like "scmRef" should really be Component Properties, but I know we don't have that implenented yet, we can do that in the next iteration.

**Review by smlambert — APPROVED** [review]
thanks @SehrishHussain and @andrew-m-leonard !

**Pull request merged**